Open ai with Locked down bot azd template

graemefoster opened the pull request:
Please fill out this template! There are three different types of contributions, feel free to delete the checklists that are not applicable to your contribution type. 

# If you are submitting a new `azd` template to the gallery
Fill this out if you want your template to be added to the awesome-azd gallery!

<!-- Please ensure that your PR includes the following metadata! All fields are mandatory unless explicitly marked as optional -->

- [ ] Added an entry to https://github.com/Azure/awesome-azd/blob/main/website/src/data/users.tsx that includes:
    - [ ] **Template title** - A short title that reflects the local application stack that someone could use to get their application on Azure (e.g. "Containerized React Web App with Java API and MongoDB")
    - [ ] **Description** - 1-2 sentence description of the architecture (e.g. Azure services) or solution that is defined by the template.
    - [ ] **Architecture Diagram or Application Screenshot** - Used as display image for gallery card. The image should include all services and their connections ([example](https://github.com/Azure-Samples/todo-csharp-sql/blob/main/assets/resources.png)). You should add the image to the `website/src/data/images/`.
    - [ ] **Link to Author's GitHub or other relevant website** - Used for attribution
    - [ ] **Author's Name** -  Name to credit on the gallery card
    - [ ] **Link to template source** - Link to the template GitHub repo 
    - [ ] **Tags** - One or more [tags](https://github.com/Azure/awesome-azd/blob/main/website/src/data/tags.tsx) representing the template. Provide at least 1 tag for programming language used and at least 1 tag for Azure services integrated. Also tag the IaC provider (Bicep or Terraform). If you don't see a relevant tag for your template? Feel free to add one!

- [ ] In the PR comment, if you can also add a link to the PR where you made your repo `azd` compatible this will allow us to provide feedback on your template and speed up the review process.

# If you are submitting a resource to be added to the awesome-azd README:
- [] Name of resource
- [] Resource author - who created this resource? (so we can credit them!)
- [] What section should this resource be included in? -Is the resource an article? A video? Something else?

<!-- Once submitted, the issue will be reviewed - we plan to do reviews on a rolling basis at regular intervals. The process will include verifying all information required for the template gallery is provided and the template works (i.e., successfully deploys to Azure with `azd up`). 
 * If we have questions or enhancements, we will add comments in issue thread (**issue stays open**)
 * If the contribution is approved, we'll merge the PR to update the gallery (**issue will then be closed**) -->

v-xuto commented:
> @v-xuto Would you test this template when you get a chance? 

Sure. We will start testing this template today.

v-xuto commented:
@hemarina We have finished the template test, and we have filed two issues. Please review.
- https://github.com/graemefoster/LockedDownChatBot/issues/1
- https://github.com/graemefoster/LockedDownChatBot/issues/2

graemefoster commented:
> @hemarina We have finished the template test, and we have filed two issues. Please review.
> 
> * [System.ArgumentNullException: Value cannot be null graemefoster/LockedDownChatBot#1](https://github.com/graemefoster/LockedDownChatBot/issues/1)
> * [Use `azd init` instead of `gh repo clone`  graemefoster/LockedDownChatBot#2](https://github.com/graemefoster/LockedDownChatBot/issues/2)

Great - let me fix these up. Thanks for the initial attempt and apologies it didn't deploy.

graemefoster commented:
@v-xuto - should be good to test again. I've updated my repo's readme.md file to show the various ways to use this template (controlled by azd env variables).

graemefoster commented on website/src/data/users.tsx:
```diff
@@ -584,6 +584,17 @@ const Users: User[] = [
+    website: 'https://github.com/graemefoster',
+    author: 'Graeme Foster',
+    source: 'https://github.com/graemefoster/LockedDownChatBot',
+    tags: ['openai', 'gpt', 'dotnet', 'appservice', 'bot', 'teams', 'cosmos', 'private']
```
@hemarina - have changed the tags (I added an azure one for the Bot service which I reference).

v-xuto commented:
> @v-xuto - should be good to test again. I've updated my repo's readme.md file to show the various ways to use this template (controlled by azd env variables).

@graemefoster We have retested this template, we found one issue. Please help review and resolve.
- https://github.com/graemefoster/LockedDownChatBot/issues/3

graemefoster commented:
> * Failed to create openai service graemefoster/LockedDownChatBot#3

Thanks @v-xuto . This looks like a capacity problem on your subscription rather than a template problem so I don't think I can do anything to help here.... See https://github.com/Azure-Samples/azure-search-openai-demo/issues/313 for a similar problem. 

I've lowered the default capacity on mine to 20, but have you purged the other OpenAI services you've created from other templates? FWIW this one (https://github.com/pascalvanderheiden/ais-apim-openai) just got approved with pretty much the same bicep mine has.

Hopefully this helps you get mine merged in.

v-jiaodi commented:
@graemefoster For https://github.com/graemefoster/LockedDownChatBot/issues/3, I had added new [comments](https://github.com/graemefoster/LockedDownChatBot/issues/3#issuecomment-1665418785).

graemefoster commented:
> @graemefoster For [graemefoster/LockedDownChatBot#3](https://github.com/graemefoster/LockedDownChatBot/issues/3), I had added new [comments](https://github.com/graemefoster/LockedDownChatBot/issues/3#issuecomment-1665418785).

Thanks @v-xuto. I've pushed a new version that allows you to specify OpenAi location using the env var "OPENAI_LOCATION", and added doco for it. I've kept the default as 'eastus'. Problem I have is that OpenAI isn't available in most locations, so you have to choose a default at the very least. 

Might be worth checking every template that creates OpenAI can deploy when you choose locations like australiaeast that don't have the service. I suspect many of the ones in the repo will break... For example, this one approved last week (https://github.com/pascalvanderheiden/ais-apim-openai/blob/main/infra/main.bicep) will blow up if you set location to one that OpenAI isn't available in.

Thanks,

Graeme

v-xuto commented:
@hemarina Two issues have been fixed, currently this template has no issues.

jongio commented:
Not required, but it would be great if you could use the azd core infra that we use in other azd templates.  There's some good modules and code in yours that would be good to share with other azd users as well.  You could replace a lot of the bicep code in your template as well.  You can find them here: https://github.com/Azure/azure-dev/tree/main/templates/common/infra/bicep

jongio commented:
In order for a user to use an existing resource in a different location you'll need to expose a location param for each resource group.  You can see how we do that in this template here: https://github.com/Azure-Samples/azure-search-openai-demo-csharp/blob/main/infra/main.bicep#L37

graemefoster commented:
> In order for a user to use an existing resource in a different location you'll need to expose a location param for each resource group. You can see how we do that in this template here: https://github.com/Azure-Samples/azure-search-openai-demo-csharp/blob/main/infra/main.bicep#L37

Are you sure about this @jongio? The template as-is works fine with an existing OpenAI in a different resource-group / location. (Whilst developing I had all new resources in AU-East, and OpenAI in EastUS).

The template you linked to doesn't reference an existing resource - it creates a new one in a different location.

graemefoster commented:
> Not required, but it would be great if you could use the azd core infra that we use in other azd templates. There's some good modules and code in yours that would be good to share with other azd users as well. You could replace a lot of the bicep code in your template as well. You can find them here: https://github.com/Azure/azure-dev/tree/main/templates/common/infra/bicep

I'll see what I can re-use. Would be great to package these into a container so we could reference them. They look to change quite often.

graemefoster commented:
> Not required, but it would be great if you could use the azd core infra that we use in other azd templates. There's some good modules and code in yours that would be good to share with other azd users as well. You could replace a lot of the bicep code in your template as well. You can find them here: https://github.com/Azure/azure-dev/tree/main/templates/common/infra/bicep

This might take a bit more time... Unfortunately what Bicep gives with one hand, it takes with the other... All my RBAC would need to move into the OpenAI module, or into another module (happy to do this but the loosey-goosey object loses all type safety, and I've had issues with the strong-typing which is still preview)

![image](https://github.com/Azure/awesome-azd/assets/160201/7b66c0b9-9bf0-4c50-ab45-4e424d34e1f9)
